model/src/com/redhat/ceylon/model/typechecker/model/Type.java

```diff
@@ -217,11 +217,137 @@ else if (isTypeParameter()) {
+                    }
+                }
+                if (isNotGeneric(type.getDeclaration())) {
+                    return result;
+                }
+            }
+        }
+        
+        if (isClassOrInterface()) {
+            if (type.isTypeParameter()) {
+                return false;
+            }
+            if (isExactlyNothing()) {
+                return type.isExactlyNothing();
+            }
+            if (type.isUnion()) {
+                boolean result = false;
+                for (Type ct: type.getCaseTypes()) {
+                    if (isExactly(ct)) {
+                        result = true;
+                    }
+                    else if (!isSubtypeOf(ct)) {
+                        return false;
+                    }
+                }
+                return result;
+            }
+            if (type.isIntersection()) {
+                boolean result = false;
+                for (Type st: type.getSatisfiedTypes()) {
+                    if (isExactly(st)) {
+                        result = true;
+                    }
+                    else if (!isSupertypeOf(st)) {
+                        return false;
+                    }
+                }
+                if (isNotGeneric(getDeclaration())) {
+                    return result;
+                }
+            }
+        }
+        
+        if (isClassOrInterface() 
+                && type.isClassOrInterface()) {
+            //already handled above!
+            /*if (isExactlyNothing()) {
+                return type.isExactlyNothing();
+            }
+            if (type.isExactlyNothing()) {
+                return isExactlyNothing();
+            }*/
+            TypeDeclaration dec = 
+                    eliminateAlias(getDeclaration());
+            TypeDeclaration otherDec = 
+                    eliminateAlias(type.getDeclaration());
+            if (!otherDec.equals(dec)) {
+                return false;
+            }
+            if (isNotGeneric(dec)) {
+                return true;
+            }
+        }
+        
+        if (isTypeParameter() 
+                && type.isTypeParameter()) {
+            TypeDeclaration dec = 
+                    getDeclaration();
+            TypeDeclaration otherDec = 
+                    type.getDeclaration();
+            if (!otherDec.equals(dec)) {
+                return false;
+            }
+            if (isNotGeneric(dec)) {
+                return true;
+            }
+        }
+        
+        //otherwise we need to resolve aliases
+        //and canonicalize the types
         return type!=null && 
                 resolveAliases()
                     .isExactlyInternal(
                             type.resolveAliases());
     }
+
+    private static boolean isNotGeneric(TypeDeclaration dec) {
+        if (!dec.getTypeParameters().isEmpty()) {
+            return false;
+        }
+        if (dec.isClassOrInterfaceMember()
+                && !dec.isStatic()) {
```

Comment:
Is the `!dec.isStatic()` test right? Static "members" can still see their container's TPs